### ISSUE TITLE
[Parallelization] Design API

### ISSUE THREAD

**Issue by pabloferz** [SSAGESLabs/PySAGES#54]
Not clear how the API would look like. Any suggestions here are welcome.

**Comment by pabloferz** [comment]
@InnocentBug, @johnaparker

**Comment by johnaparker** [comment]
For parallel execution to work, the MD execution has to moved to a global function, where the inputs and outputs are pickle-able objects. One way to accomplish this is to pass `sampler_args` and `context_args` to generate the Sampler and Context objects, e.g., for Umbrella Integration:
```python
# Global definition
def run_umbrella_sampling(sampler_args, context_generator, context_args, center, kspring, hist_period, hist_offset, timesteps):
    sampler = UmbrellaIntegration(**sampler_args)
    sampler.center = center
    sampler.kspring = kspring
    callback = HistogramLogger(hist_period, hist_offset)

    context = context_generator(**context_args)
    wrapped_context = ContextWrapper(context, sampler, callback)

    with wrapped_context:
        wrapped_context.run(timesteps)

    mean = callback.get_means()
    nabla_A = free_energy_gradient(sampler.kspring, mean, center)
    return dict(kspring=kspring, center=center, histogram=callback, histogram_means=mean, nabla_A=nabla_A)
```

This function can then be dispatched to some parallel scheme, e.g., Dask or MPI Pool:
```python
futures = []
with MPIPoolExecutor() as executor:
    for rep in range(Nreplica):
        context_args["replica_num"] = rep
        sampler_args = dict(cvs=self.cvs)
        futures.append(executor.submit(run_umbrella_sampling, sampler_args, context_generator, context_args, centers[rep], ksprings[rep], hist_periods[rep], hist_offsets[rep], timesteps[rep]))
    for future in futures:
        for key,val in future.result().items():
            result[key].append(val)
```
This approach is implemented in 0d8be39

We'll need to think about if this is the cleanest way to do this. It's a little strange that the sampler object's `run` function is itself spawning sampler objects.

**Comment by InnocentBug** [comment]
The original idea that we had with the context generation was exactly to avoid this situation.
Just have a function that generates every non-picklable object.

Maybe as an idea we abstract the sampling objects and CVS a bit more.
Such that we have a user facing class, that behaves as before, but doesn't acutally initialize anything with the backend.
(So basically just a shell, that stores the arguments the user provided.)
And a second one, that can be generate the actual instances at the point of execution after deployment, by fetching the arguments fro the first one.

Just a rough idea, that maybe allows us to keep the structure somewhat similar.
